src/java.base/share/classes/com/sun/crypto/provider/GaloisCounterMode.java

```diff
@@ -198,820 +542,1177 @@ private void checkDataLength(int ... lengths) {
+            checkReInit();
+            byte[] output = new byte[getOutputSize(inLen, false)];
+            try {
+                doUpdate(in, inOff, inLen, output, 0);
+            } catch (ShortBufferException e) {
+                // update decryption has no output
+            }
+            return output;
         }
 
-        byte[] block = getLengthBlock(sizeOfAAD, processed);
-        ghashAllToS.update(block);
-        block = ghashAllToS.digest();
-        GCTR gctrForSToTag = new GCTR(embeddedCipher, this.preCounterBlock);
-        gctrForSToTag.doFinal(block, 0, tagLenBytes, block, 0);
+        /**
+         * Encrypt update operation.  This uses both the ibuffer and 'in' to
+         * encrypt as many blocksize data as possible.  Any remaining data is
+         * put into the ibuffer.
+         */
+        @Override
+        public int doUpdate(byte[] in, int inOfs, int inLen, byte[] out,
+            int outOfs) throws ShortBufferException {
+            checkReInit();
+            checkDataLength(inLen, getBufferedLength());
+            ArrayUtil.nullAndBoundsCheck(in, inOfs, inLen);
+            ArrayUtil.nullAndBoundsCheck(out, outOfs, out.length - outOfs);
+
+            // 'inLen' stores the length to use with buffer 'in'.
+            // 'len' stores the length returned by the method.
+            int len = 0;
+            int bLen = getBufferedLength();
+
+            processAAD();
+            out = overlapDetection(in, inOfs, out, outOfs);
+
+            // if there is enough data in the ibuffer and 'in', encrypt it.
+            if (bLen > 0) {
+                byte[] buffer = ibuffer.toByteArray();
+                // number of bytes not filling a block
+                int remainder = bLen % blockSize;
+                // number of bytes along block boundary
+                bLen -= remainder;
+
+                // If there is enough bytes in ibuffer for a block or more,
+                // encrypt that first.
+                if (bLen > 0) {
+                    len += gctrghash.update(buffer, 0, bLen, out, outOfs);
+                    outOfs += bLen;
+                }
 
-        System.arraycopy(block, 0, out, (outOfs + len), tagLenBytes);
-        return (len + tagLenBytes);
-    }
+                // blen is now the offset for 'buffer'
+
+                // Construct and encrypt a block if there is enough 'buffer' and
+                // 'in' to make one
+                if ((inLen + remainder) >= blockSize) {
+                    byte[] block = new byte[blockSize];
+
+                    System.arraycopy(buffer, bLen, block, 0, remainder);
+                    int inLenUsed = blockSize - remainder;
+                    System.arraycopy(in, inOfs, block, remainder, inLenUsed);
+
+                    len += gctrghash.update(block, 0, blockSize, out, outOfs);
+                    inOfs += inLenUsed;
+                    inLen -= inLenUsed;
+                    outOfs += blockSize;
```

Comment:
Either is fine because len and blockSize will be the same.

Reply:
Hmm, re-reading the code, I see why you chose to use blockSize since line 1055 uses "len += ..." so whether len==blockSize depends on the code in line 1022-1054.